fairscale/optim/oss.py

```diff
@@ -319,12 +306,129 @@ def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
+        """
+        return self.optim.state_dict()
+
+    def state_dict(self) -> Dict[str, Any]:
+        """Return the last known global optimizer state. The returned state is compatible with Pytorch, in that the
+        sharded properties are not exposed. It contains two entries:
+
+        * state - a dict holding current optimization state. Its content
+            differs between optimizer classes.
+
+        * param_groups - a dict containing all parameter groups
+
+        .. warning:
+            If the state has not been consolidated, this returns a shard's worth, not the global state.
+
+        .. warning:
+            Returning the global state is limited to the replica which was responsible for the consolidation.
+            The state may also not be up to date, depending on when `consolidate_state_dict` was last called.
+        """
+
+        if len(self._all_states) == 0:
+            raise RuntimeError(
+                "Optimizer state has not been consolidated on this rank. \
+                Please call `consolidate_state_dict()` on all ranks beforehand if you meant to save the global state"
+            )
+
+        # Unify the shard states and the state that pytorch would expect, given the model.
+        # Indexation needs several redirections, since each shard only knows a limited scope of the model
+        # - get the pytorch compliant parameter indexing
+        state_dict = super().state_dict()
+
+        # - get an id map which links the parameter id to the index in the reference state
+        global_id_map = {id(p): i for i, p in enumerate(chain(*(g["params"] for g in self.param_groups)))}
+
+        # - go through the per-shard states, which are all indexed locally
+        unordered_state = {}
+        for rank, s in enumerate(self._all_states):
+            # -- match the local indexing and the global partition, update the corresponding saved state globally
+            for local_pg, global_pg in zip(s["param_groups"], self.partition_parameters()[rank]):
+                # Go through the parameters indexed locally, pick up the global corresponding param
+                # NOTE: Contents of the state_dict changes in between torch1.5 and torch1.6+
+                local_index_to_param_id = {
+                    i_param: id(global_pg["params"][i]) for i, i_param in enumerate(local_pg["params"])
+                }
+
+                for local_param_index in local_pg["params"]:
+                    # Update the state, if any
+                    if local_param_index in s["state"].keys():
+                        unordered_state[local_index_to_param_id[local_param_index]] = s["state"][local_param_index]
+
+        # - save the states in the expected order, meaning the one from the param_groups
+        for i, param in enumerate(chain(*(g["params"] for g in self.param_groups))):
+            if id(param) in unordered_state.keys():
+                state_dict["state"][i] = unordered_state[id(param)]
```

Comment:
Why can't we just add the parameter directly here to the returned state_dict and avoid the later loops?
```suggestion
                for local_param, global_param in zip(local_pg["params"], global_pg["params"]):
                    global_id = global_id_map[id(global_param)]
                    state_dict["state"][global_id] = local_parm
```

Reply:
kind of a mind breaker (the indices all flip) but I think that can work indeed, cleaner, thanks for the suggestion

Reply:
just to explain a little, the suggestion does not work out of the box because the `local_pg["params"]` is just a list of indices, there are no real params in there, and we need to pull the corresponding state from the smaller optimizer, but I think that the idea works in that I can reassign directly here instead of looping twice indeed

Reply:
[done, not exact same as explained above but similar enough, one less loop]